File: problems/198/198.house-robber.py
#
# @lc app=leetcode id=198 lang=python3
#
# [198] House Robber
#

# @lc code=start
class Solution:
    # def rob(self, nums: List[int]) -> int:
    #     # bottom up dp

    #     n = len(nums)
    #     dp = [0] * (n + 1)
    #     dp[1] = nums[0]
        
    #     for i in range(1, n):
    #         dp[i + 1] = max(dp[i - 1] + nums[i], dp[i])
        
    #     return dp[n]

    def rob(self, nums: List[int]) -> int:
        # bottom up dp with O(1) space
        prev1, prev2 = 0, 0
        
        for i in range(len(nums)):
            temp = max(prev1 + nums[i], prev2)
            prev1 = prev2
            prev2 = temp
        
        return prev2
        
# @lc code=end

File: problems/198/198.house-robber.java
/*
 * @lc app=leetcode id=198 lang=java
 *
 * [198] House Robber
 */

// @lc code=start
class Solution {
    public int rob(int[] nums) {
        /*
         * cannot rob 2 adjacent houses.
         * return max amount of money that can be robbed
         *
         * intuition:
         * - dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])
         * - depends only on prev 2
         *
         * O(N) time, O(1) space
         */

        if (nums.length == 1) {
            return nums[0];
        }

        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }

        int prev = nums[0];
        int curr = Math.max(nums[0], nums[1]);

        for (int i = 2; i < nums.length; i++) {
            int temp = curr;
            curr = Math.max(curr, prev + nums[i]);
            prev = temp;
        }

        return curr;
    }
}
// @lc code=end
